src/passes/J2CLOpts.cpp

```diff
@@ -30,9 +30,48 @@ namespace wasm {
 
 namespace {
 
+using AssignmentCountMap = std::unordered_map<Name, Index>;
+using TrivialFunctionMap = std::unordered_map<Name, Expression*>;
+
 bool isOnceFunction(Function* f) { return f->name.hasSubstring("_<once>_"); }
 
-using AssignmentCountMap = std::unordered_map<Name, Index>;
+// Returns the function body if it is a trivial function, null otherwise.
+Expression* getTrivialFunctionBody(Function* func) {
+  auto* body = func->body;
+
+  // Only consider trivial the following instructions which can be safely
+  // inlined and note that their size is at most 2.
+  if (body->is<Nop>() || body->is<GlobalGet>() || body->is<Const>() ||
+      (body->is<Call>() && body->dynCast<Call>()->operands.size() == 0) ||
+      (body->is<GlobalSet>() &&
+       body->dynCast<GlobalSet>()->value->is<Const>())) {
+    return body;
+  }
+  return nullptr;
+}
+
+// Adds the function to the map if it is trivial.
+void maybeCollectTrivialFunction(Module* module,
```

Comment:
we need to always pair maybeCollectTrivialFunction with a cleanupFunction so let's have a function that abstracts the two.

Reply:
See my comment above. The collector does not seem to need cleanup.